src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,226 @@
+  (A + B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.sub [add_group α]
+  {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+@[simp] lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
+  {k : R} {A : matrix n n α} (ha : A.is_diag) :
+  (k • A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+@[simp] lemma is_diag.transpose
+  [has_zero α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᵀ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.conj_transpose
+  [semiring α] [star_ring α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᴴ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.minor [has_zero α]
+  {A : matrix n n α} (ha : A.is_diag) {f : m → n} (hf : injective f) :
+  (A.minor f f).is_diag :=
+by { intros i j h, specialize @hf i j, simp [ha (mt hf h)] }
+
+/-- `(A ⊗ B).is_diag` if both `A` and `B` are diagonal. -/
+@[simp] lemma is_diag.kronecker [mul_zero_class α]
+  {A : matrix m m α} {B : matrix n n α} (hA: A.is_diag) (hB: B.is_diag) :
+  (A ⊗ₖ B).is_diag :=
+begin
+  rintros ⟨a, b⟩ ⟨c, d⟩ h,
+  dsimp [kronecker_apply],
+  by_cases hac: a = c,
+  { have hbd: b ≠ d, {tidy}, simp [hB hbd] },
+  { simp [hA hac] },
+end
+
+@[simp] lemma is_diag.diagonal [has_zero α] [decidable_eq n] (d : n → α) :
+  (diagonal d).is_diag :=
+by { intros i j h, simp [h] }
+
+lemma is_diag.is_symm [has_zero α] {A : matrix n n α} (h : A.is_diag) :
+  A.is_symm :=
+begin
+  ext,
+  by_cases g : i = j, {rw g},
+  simp [h g, h (ne.symm g)],
+end
+
+/-- The block matrix `A.from_blocks 0 0 D` is diagonal if  `A` and `D` are diagonal. -/
+lemma is_diag.from_blocks [has_zero α]
+  {A : matrix m m α} {D : matrix n n α}
+  (ha : A.is_diag) (hd : D.is_diag) :
+  (A.from_blocks 0 0 D).is_diag :=
+begin
+  rintros (i | i) (j | j) hij,
+  { have h' := ne_of_apply_ne (λ (i : m), sum.inl i) hij, simp [ha h'] },
+  { simp * at * },
+  { simp * at * },
+  { have h' := ne_of_apply_ne (λ (i : n), sum.inr i) hij, simp [hd h'] },
+end
+
+/-- This is the `iff` version of `matrix.is_diag.from_blocks`. -/
+lemma is_diag_from_blocks_iff [has_zero α]
+  {A : matrix m m α} {B : matrix m n α} {C : matrix n m α} {D : matrix n n α} :
+  (A.from_blocks B C D).is_diag ↔ A.is_diag ∧ B = 0 ∧ C = 0 ∧ D.is_diag :=
+begin
+  split,
+  { intros h,
+    repeat {split};
+    try {intros i j hij <|> ext i j},
+    { have g : sum.inl i ≠ sum.inl j, {simp [hij]},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inl i ≠ sum.inr j, {simp},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inr i ≠ sum.inl j, {simp},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inr i ≠ sum.inr j, {simp [hij]},
+      specialize h g,
+      simp [from_blocks, *] at * } },
+  { rintros ⟨ha, hb, hc, hd⟩,
+    convert is_diag.from_blocks ha hd }
+end
```

Comment:
This is a much shorter proof:
```
begin
  split,
  { intros h,
    refine ⟨_, _, _, _⟩,
    { intros i j hij,
      exact h (sum.inl_injective.ne hij), },
    { ext i j,
      exact h sum.inl_ne_inr, },
    { ext i j,
      exact h sum.inr_ne_inl, },
    { intros i j hij,
      exact h (sum.inr_injective.ne hij), }, },
  { rintros ⟨ha, hb, hc, hd⟩,
    convert is_diag.from_blocks ha hd }
end
```
For trivial goals, you should usually try `library_search` or `squeeze_simp` before `simp`.